### PR TITLE
avoid appending directory as file path in s3 driver Walk

### PR DESCRIPTION
**Pull request by flavianmissi**
when a directory is empty, the s3 api lists it with a trailing slash.
this causes the path to be appended twice to the walkInfo slice, causing
purge uploads path transformations to panic when the `_uploads` is
emtpy.

this adds a check for file paths ending on slash, and do not append
those as regular files to the walkInfo slice.

fixes https://github.com/distribution/distribution/issues/4358

### CODE REVIEW

**Review by milosgajdos — COMMENTED** [review]
I need to have a proper look at this, but this seems like a bit of a "hack" 🤔

**Comment by flavianmissi** [comment]
> I need to have a proper look at this, but this seems like a bit of a "hack" 🤔

Yeah that's fair. I'll dig some more and see if there's a better way to solve this. 
This [comment](https://github.com/distribution/distribution/blob/main/registry/storage/driver/s3-aws/s3.go?plain=1#L1154-L1157) gives interesting insight into the current behavior:
```
// When the "delimiter" argument is omitted, the S3 list API will list all objects in the bucket
// recursively, omitting directory paths. Objects are listed in sorted, depth-first order so we
// can infer all the directories by comparing each object path to the last one we saw.
// See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html
```

The claim that directory paths are omitted is only true when the directory in question is not empty. `directoryDiff` seems to only work on non-empty directories.

Here's a sample output of list objects v2:
```
$ aws s3api list-objects-v2 --bucket fmissi-images --prefix=s3walktest
CONTENTS        "f339720b172fa37468491c1d298a2f70"      s3walktest/docker/registry/v2/blobs/sha256/04/046909    2024-10-15T08:38:29+00:00       50      STANDARD
CONTENTS        "47f2f2a00ed0450266d0385769ef0347"      s3walktest/docker/registry/v2/blobs/sha256/07/071a45    2024-10-15T08:38:29+00:00       50      STANDARD
CONTENTS        "44de784b8d8e7a1b9af85dffb5f072f2"      s3walktest/docker/registry/v2/repositories/testns/testimg/_layers/sha256/2a43dc 2024-10-15T08:38:30+00:00       77      STANDARD
CONTENTS        "27516816763353e1abfd9501a319a4f2"      s3walktest/docker/registry/v2/repositories/testns/testimg/_layers/sha256/3ae7e8 2024-10-15T08:38:30+00:00       77      STANDARD
CONTENTS        "eed715501bf8c634e10462dff0375db4"      s3walktest/docker/registry/v2/repositories/testns/testimg/_manifests/revisions/sha256/3ae7e8    2024-10-15T08:38:30+00:00       90      STANDARD
CONTENTS        "1fc2b1056c14903607fc17c961f90ee3"      s3walktest/docker/registry/v2/repositories/testns/testimg/_uploads/     2024-10-15T08:38:30+00:00       65      STANDARD
```

**Comment by milosgajdos** [comment]
> The claim that directory paths are omitted is only true when the directory in question is not empty. directoryDiff seems to only work on non-empty directories.

I have often wondered about that comment, though I can't remember if I ever looked into whether those statements hold 🥲  

Ah them pesky edge conditions.........I'd still prefer modifying `directoryDiff` 🤔

**Comment by flavianmissi** [comment]
> Ah them pesky edge conditions.........I'd still prefer modifying `directoryDiff` 🤔

After looking into it I think `directoryDiff` is innocent. It correctly returns the `_uploads` entry as a dir.
I wrote the following test for it:
```
func TestDirectoryDiff(t *testing.T) {
	prefix := "docker/registry/v2"
	filePath := "docker/registry/v2/repositories/testns/testimg/_uploads/"
	dirs := directoryDiff(prefix, filePath)
	expectedDirs := []string{
		"docker/registry/v2/repositories",
		"docker/registry/v2/repositories/testns",
		"docker/registry/v2/repositories/testns/testimg",
		"docker/registry/v2/repositories/testns/testimg/_uploads",
	}
	if !slices.Equal(dirs, expectedDirs) {
		t.Fatalf("directoryDiff returned incorrect list of dirs.\nwanted: %v\ngot:    %v", expectedDirs, dirs)
	}
}
```
It works as expected:
```
go test -v -run TestDirectoryDiff  ./registry/storage/driver/s3-aws
=== RUN   TestDirectoryDiff
--- PASS: TestDirectoryDiff (0.00s)
PASS
ok  	github.com/distribution/distribution/v3/registry/storage/driver/s3-aws	1.033s
```

Could you expand on how you would expect `directoryDiff` to behave instead? 
 
The entire loop through `objects.Contents` in `doWalk` was written with the premise that every `file.Key` entry listed is an actual file path, and we just found out that this isn't the case. I'm honestly not sure how to fix this without treating it like an exception (and without rewriting the entire loop, which I can do but didn't think necessary).

Anyways, I'll pause this while I wait for further feedback. 
Thanks for reviewing o/

**Comment by flavianmissi** [comment]
Ohh, thanks for clarifying. 
Either way I'll sleep on it, maybe I'll wake up with an idea of a more elegant way to approach this.

**Comment by flavianmissi** [comment]
oh my look at what I found: https://github.com/distribution/distribution/commit/87cbd09fa7ffbdc1b8eb2ab5098a5e91e815c216#diff-69ca1a0abf51616b5031c87b33ce5a35fb6439b5bcd744aef65ddbc2131ec431R170
this explains why older versions do not panic 😞

**Comment by milosgajdos** [comment]
> this explains why older versions do not panic 😞

Removing a test explains why the older version doesn't panic? Very strange and kinda.... worrying

**Comment by milosgajdos** [comment]
oh you mean this https://github.com/distribution/distribution/pull/3302

**Comment by flavianmissi** [comment]
Yes! [PR 3302](https://github.com/distribution/distribution/pull/3302) fixes the same error reported on https://github.com/distribution/distribution/issues/4358, and the regression is caused by https://github.com/distribution/distribution/commit/f78d81e78a20fda60efb4a29848f77d99dd0513b which not only removes the test but also the code treating the /.

**Comment by milosgajdos** [comment]
Ok, makes sense. Thanks for digging in. Let's roll with this then!

File: registry/storage/driver/s3-aws/s3.go [hunk 1]
```diff
@@ -1178,6 +1178,16 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, from stri
 				prevDir = dir
 			}
 
+			// in some cases the _uploads dir might be empty. when this happens, it would
+			// be appended twice to the walkInfos slice, once as [...]/_uploads and
+			// once more erroneously as [...]/_uploads/. the easiest way to avoid this is
+			// to skip appending filePath to walkInfos if it ends in "/". the loop through
+			// dirs will already have handled it in that case, so it's safe to continue this
+			// loop.
+			if strings.HasSuffix(filePath, "/") {
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1187]: Ignore this wondering, see: https://github.com/distribution/distribution/pull/4485#issuecomment-2417088354

**Review by milosgajdos — APPROVED** [review]
PTAL @Jamstah

File: registry/storage/driver/s3-aws/s3_test.go [hunk 2]
```diff
@@ -514,6 +514,86 @@ func TestDelete(t *testing.T) {
...
+		}
+		resp, err := d.S3.DeleteObjectsWithContext(ctx, &s3.DeleteObjectsInput{
+			Bucket: aws.String(d.Bucket),
+			Delete: &s3.Delete{
+				Objects: s3Objects,
+				Quiet:   aws.Bool(false),
+			},
+		})
+		if err != nil {
+			t.Logf("DeleteObjectsWithContext resp: %+v", resp)
```
Comment (squizzi) [inline registry/storage/driver/s3-aws/s3_test.go:577]: thanks for these extras, they'll be useful for test debugging to someone one day

**Review by squizzi — APPROVED** [review]
lgtm, @Jamstah should probably take a look too though
